platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.roster;
+
+import com.hedera.hapi.node.base.ServiceEndpoint;
+import com.hedera.hapi.node.state.primitives.ProtoBytes;
+import com.hedera.hapi.node.state.roster.Roster;
+import com.hedera.hapi.node.state.roster.RosterEntry;
+import com.hedera.hapi.node.state.roster.RosterState;
+import com.hedera.hapi.node.state.roster.RoundRosterPair;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.base.utility.Pair;
+import com.swirlds.platform.state.MerkleRoot;
+import com.swirlds.platform.system.address.AddressBook;
+import com.swirlds.state.State;
+import com.swirlds.state.spi.ReadableKVState;
+import com.swirlds.state.spi.ReadableSingletonState;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.security.cert.CertificateEncodingException;
+import java.util.Comparator;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * A utility class to help retrieve a Roster instance from the state.
+ */
+public final class RosterRetriever {
+    private RosterRetriever() {}
+
+    private static final String IP_ADDRESS_COMPONENT_REGEX = "(\\d{1,2}|(?:0|1)\\d{2}|2[0-4]\\d|25[0-5])";
+    private static final Pattern IP_ADDRESS_PATTERN =
+            Pattern.compile("^%N\\.%N\\.%N\\.%N$".replace("%N", IP_ADDRESS_COMPONENT_REGEX));
+
+    /**
+     * Retrieve the current active Roster from the state.
+     * <p>
+     * This method first checks the RosterState/RosterMap entities,
+     * and if they contain an active roster, then returns it.
+     * If the active roster is missing from RosterState,
+     * then fall back to reading an AddressBook from the PlatformState
+     * and converting it to a Roster.
+     *
+     * @return an active Roster for the round of the state, or a Roster that represents the current AddressBook in PlatformState
+     */
+    @NonNull
+    public static Roster retrieve(@NonNull final State state) {
+        final Bytes activeRosterHash = getActiveRosterHash(state);
+        if (activeRosterHash != null) {
+            final ReadableKVState<ProtoBytes, Roster> rosterMap =
+                    state.getReadableStates("RosterServiceImpl").get("ROSTER");
```

Comment:
I considered that originally. However, this would introduce a dependency of the platform core module on an app module. I don't know whether a circular dependency will be created this way, but even if it won't, it didn't feel right to make the platform core depend on the app.

Do you think it's a good idea?

Reply:
It should be fine to move it from the schema and service into a class in some module that's common to both.

Reply:
Happy to do that. Could you please suggest a place (a specific class, or at least a java package/module)?

Reply:
swirlds-common is a module common to both, so i think that's a good place (except someone else has a strong reason for not putting it there)

Reply:
Cool. Done.